intellij/src/saros/intellij/ui/actions/ConnectServerAction.java

```diff
@@ -51,37 +60,131 @@ public void execute() {
+
+      return;
+    }
+
+    if (newJID.equals(currentJID)) {
+      LOG.debug(
+          "Ignoring connection action as account is already connected. current JID: "
+              + currentJID
+              + ", new JID: "
+              + newJID);
+
+      return;
+    }
+
+    ISarosSession sarosSession = sarosSessionManager.getSession();
+
+    if (sarosSession == null) {
+      connectToAccount(account, true, false);
+
+      return;
+    }
+
+    String message = Messages.ConnectServerAction_leave_session_confirmation_message;
+
+    if (sarosSession.isHost()) {
+      message += Messages.ConnectServerAction_leave_session_confirmation_host_addendum_message;
+    }
+
+    boolean reallyReconnect;
+
+    try {
+      reallyReconnect =
+          SafeDialogUtils.showYesNoDialog(
+              project, message, Messages.ConnectServerAction_leave_session_confirmation_title);
+
+    } catch (IllegalAWTContextException e) {
+      LOG.error("Failed to show question on whether to really switch accounts to user", e);
+
+      NotificationPanel.showError(
+          MessageFormat.format(
+              Messages.ConnectServerAction_leave_session_confirmation_host_addendum_error_message,
+              e.getMessage()),
+          Messages.ConnectServerAction_leave_session_confirmation_host_addendum_error_title);
+
+      return;
+    }
+
+    if (reallyReconnect) {
+      connectToAccount(account, true, true);
+    }
+  }
+
+  /**
+   * Connects to the given XMPP account.
+   *
+   * <p>If the information is passed that there already is an existing connection, the previous
+   * connection will be ended before connecting to the new account.
+   *
+   * <p>If the information is passed that there already is a running session, it will also be left
+   * before disconnecting from the currently connected XMPP account.
    *
    * @param account the account to connect to
+   * @param isAlreadyConnected whether there currently is an existing connection
+   * @param isInSession whether there currently is a running session
    */
-  private void connectAccount(final XMPPAccount account) {
-    ProgressManager.getInstance()
-        .run(
-            new Task.Modal(project, "Connecting...", false) {
-
-              @Override
-              public void run(@NotNull ProgressIndicator indicator) {
-
-                LOG.info(
-                    "Connecting server: ["
-                        + account.getUsername()
-                        + "@"
-                        + account.getServer()
-                        + "]");
-
-                indicator.setIndeterminate(true);
-
-                try {
-                  connectionHandler.connect(account, false);
-                } finally {
-                  indicator.stop();
-                }
-              }
-            });
+  private void connectToAccount(
+      @NotNull XMPPAccount account, boolean isAlreadyConnected, boolean isInSession) {
+
+    ThreadUtils.runSafeAsync(
```

Comment:
Seems like ```runProcessWithProgressSynchronously``` is the method you are looking for.

Reply:
The execution of a modal task in l.170 already ensures that the runnable is run synchronously on the EDT. And that exactly is the issue. Executing the session teardown synchronously on the EDT here leads to IDE lockups and general issues during the teardown:

```
ERROR 12:00:41,585 [Connecting to intellij_test_3@saros-con.imp.fu-berlin.de] (ActivityHandler.java:279) saros-activity-dispatcher thread is still running
ERROR 12:00:51,596 [Connecting to intellij_test_3@saros-con.imp.fu-berlin.de] (ConsistencyWatchdogServer.java:160) Consistency watchdog server is still running
```

This is why the current logic is running the whole thing in an asynchronous block on the EDT as I wanted to ensure that the execution of the session disconnect and the account reconnect is still always run together in sequence.

Using `ProgressManager.runProcessWithProgressSynchronously()` runs into the same issues.

Excluding the session disconnect from the EDT block does not work either, so it seem like the asynchronous execution on the EDT is necessary.

The only other option I could find was to make the task monitor `Backgroundable` using the option `DEAF` and moving the session disconnect back into the task. This causes the task to be run asynchronously while displaying a progress indicator. As a side effect, the progress monitor could then also be moved to the background, meaning a "Background" button is displayed as part of the dialog which causes the monitor to no longer be displayed. This somewhat breaks with the look of the other task monitors but it doesn't really matter in my opinion.

Reply:
I always forget that the IntelliJ stuff works a bit different. This is no problem in Eclipse as you can execute another UI runable while a modal dialog is present.

Just one question. Is the action disabled after you clicked it ? If you can still click it and the disconnect is not that fast a user might click on it multiple times which results in a non deterministic behavior.

Reply:
> Just one question. Is the action disabled after you clicked it ? If you can still click it and the disconnect is not that fast a user might click on it multiple times which results in a non deterministic behavior.

This option is available as part of the pop-up menu when clicking the "Connect" button. As soon as an account is chosen, the pop-up menu closes and, if there is currently a session, the dialog querying the user whether to really reconnect is shown. The dialog as well as the progress window are blocking, meaning the user can't access the IDE. So this should not be an issue.

Reply:
Looking at the code, inserting a Thread.Sleep() before the disconnect call should give you enough time to activate it another time through the context menu.

I remember that we have guards in Eclipse to prevent such an issue.

Reply:
The easiest solution would be to move the whole checking logic into the asynchronous EDT block. This way, even if the logic is triggered multiple times, all checks would be performed in in the right order.

Otherwise, I guess I could introduce some synchronization logic that ensures that everything is executed in the right order in such cases.

But, in general,  I am not sure whether it is really worth it to cover such an edge case. It would require the user's machine to be soooo slow that it takes the background EDT thread a user-intractable timeframe to get to the stage where the dialog or progress monitor is shown. Furthermore, the user would have to again re-open the popup menu and select another account in that timeframe.

Reply:
Also, this is probably an issue with all the buttons/button menus in Saros/I as non of them have a specific logic dealing with concurrent executions as far as I know.